File: tests/benchmarks/script-engines/addons/extensions/bench.js
// Script engine benchmark, JavaScript side. Six workloads over
// ctx.corpus.symbols, timed from inside the script so extraction and
// rendering are excluded. bench.lua runs the same six in the same order;
// run.py collects the "bench-js: <name> <ms> ms" lines from both.
mrdocs.register_transform("bench-js", function(ctx) {
    var s = ctx.corpus.symbols, t, r;
    t = Date.now(); var counts = {};
    for (var i = 0; i < s.length; ++i) { var k = s[i].kind; counts[k] = (counts[k] || 0) + 1; }
    console.log("bench-js: count-by-kind " + (Date.now() - t) + " ms");
    t = Date.now(); var names = [];
    for (var i = 0; i < s.length; ++i) { var sym = s[i]; if (sym.kind === "function") names.push(sym.name); }
    r = names.join(",").length;
    console.log("bench-js: collect-function-names " + (Date.now() - t) + " ms (" + names.length + ")");
    t = Date.now(); var out = "";
    for (var i = 0; i < s.length; ++i) { var sym = s[i]; out += sym.kind + " " + sym.name + "\n"; }
    console.log("bench-js: build-text " + (Date.now() - t) + " ms (" + out.length + " chars)");
    t = Date.now(); var deep = 0;
    for (var i = 0; i < s.length; ++i) { var sym = s[i]; var loc = sym.loc; if (loc && loc.def && loc.def.location && loc.def.location.lineNumber) deep += loc.def.location.lineNumber; }
    console.log("bench-js: nested-reads " + (Date.now() - t) + " ms (" + deep + ")");
    t = Date.now(); var acc = 0;
    for (var i = 0; i < 5000000; ++i) acc += i % 7;
    console.log("bench-js: numeric-loop " + (Date.now() - t) + " ms (" + acc + ")");
    t = Date.now(); var kept = [];
    for (var i = 0; i < s.length; ++i) kept.push(s[i]);
    console.log("bench-js: keep-all-proxies " + (Date.now() - t) + " ms (" + kept.length + ")");
});

File: tests/benchmarks/script-engines/addons/extensions/bench.lua
-- Script engine benchmark, Lua side. The same six workloads as bench.js,
-- in the same order, timed with os.clock from inside the script.
mrdocs.register_transform("bench-lua", function(ctx)
    local s = ctx.corpus.symbols
    local t = os.clock(); local counts = {}
    for _, sym in ipairs(s) do local k = sym.kind; counts[k] = (counts[k] or 0) + 1 end
    print(string.format("bench-lua: count-by-kind %.0f ms", (os.clock() - t) * 1000))
    t = os.clock(); local names = {}
    for _, sym in ipairs(s) do if sym.kind == "function" then names[#names + 1] = sym.name end end
    local r = #table.concat(names, ",")
    print(string.format("bench-lua: collect-function-names %.0f ms (%d)", (os.clock() - t) * 1000, #names))
    t = os.clock(); local parts = {}
    for _, sym in ipairs(s) do parts[#parts + 1] = sym.kind .. " " .. tostring(sym.name) .. "\n" end
    local out = table.concat(parts)
    print(string.format("bench-lua: build-text %.0f ms (%d chars)", (os.clock() - t) * 1000, #out))
    t = os.clock(); local deep = 0
    for _, sym in ipairs(s) do local loc = sym.loc; if loc and loc.def and loc.def.location and loc.def.location.lineNumber then deep = deep + loc.def.location.lineNumber end end
    print(string.format("bench-lua: nested-reads %.0f ms (%d)", (os.clock() - t) * 1000, deep))
    t = os.clock(); local acc = 0
    for i = 0, 4999999 do acc = acc + i % 7 end
    print(string.format("bench-lua: numeric-loop %.0f ms (%d)", (os.clock() - t) * 1000, acc))
    t = os.clock(); local kept = {}
    for _, sym in ipairs(s) do kept[#kept + 1] = sym end
    print(string.format("bench-lua: keep-all-proxies %.0f ms (%d)", (os.clock() - t) * 1000, #kept))
end)
